pkg/apis/serving/v1alpha1/kfservice_types.go

```diff
@@ -79,9 +91,35 @@ type SKLearnSpec struct {
 	Resources v1.ResourceRequirements `json:"resources,omitempty"`
 }
 
-// CustomSpec provides a hook for arbitrary container configuration.
-type CustomSpec struct {
-	Container v1.Container `json:"container"`
+// ExplanationSpec defines the configuration of the model explainer.
+type ExplanationSpec struct {
+	Async bool `json:"omitempty"` // Defaults false. If true, creates and outputs to a well known (named) Knative Channel
+
+	// The following fields follow a "1-of" semantic. Users must specify exactly one spec.
+	Template *TemplateSpec       `json:"template,omitempty"`
+	Alibi    *AlibiExplainerSpec `json:"alibi,omitempty"`
+	// Include other OSS out-of-the-box explainers here
+}
+
+type AlibiExplainerSpec struct {
+	// TODO(clivecox)
+}
+
+// AnalysisSpec defines the configuration of the analyzer default configuration to route traffic.
+type AnalysisSpec struct {
+	Skew    *SkewSpec    // TODO
+	Bias    *BiasSpec    // TODO
+	Logging *LoggingSpec // TODO
+	// TODO Include configuration of backend datastore. KFServing should provide an image
+}
+
+type SkewSpec struct {
```

Comment:
Should we add an OutlierSpec as well? And if so this does seem closer to ExplanationSpec where there are multiple algorithms that fit different types of models better so the user might want  control based on their model.

Reply:
I'm a little confused. Can you explain the outlier usecase? My understanding is that it's just a statistical definition?

Reply:
There are various [online outlier detection algorithms] we have been working on at Seldon(https://github.com/SeldonIO/seldon-core/tree/master/components/outlier-detection) which need to be matched to the model. So I see it as a similar case to explainability where you want to configure which you use.

Reply:
Does skew here mean for the data distribution change or shift ?

Reply:
Skew (or concept drift) would be identifying a decrease in accuracy over a time period/batch of requests while outlier detection would be for flagging individual requests as outliers to the current distribution. @jklaise @arnaudvl ?

Reply:
@cliveseldon correct, outlier detection needs to be tied to the specific example as an outlier detection is a model telling wether a particular data point is an anomaly (could be wrt to training data or wrt to samples seen so far in an online updating fashion), so it's similar to the explainer use case, while concept drift monitors the data stream as a whole and produces statistics/alarms (e.g. detecting a change point in the data stream).

Reply:
Interesting. It's almost starting to sound like outlier detection shouldn't happen within a KFService. Now I'm questioning whether or not explainability should happen within a KFService. Can we break this down into a "Whitebox" vs "Blackbox" problem, where anything that's blackbox can have OOTB solutions that enhance your model, and anything whitebox is considered custom and must be deployed outside your model.

Reply:
I think the key is more components which are online (they listen to requests/responses and build internal models) and are (optionally) part of the request/response flow - add an explanation, outlier detection alert, concept drift alert etc. They need to be associated and configured with a particular unit of prediction.

Reply:
This is where we landed thursday night. Many of the analysis components get pushed out (everything offline), and there's a small set of online components for a static graph -- somewhat similar to our original stance.